Public/Src/Engine/Processes/SharedOpaqueJournal.cs

```diff
@@ -0,0 +1,183 @@
+        /// </summary>
+        public string JournalDirectory => Path.GetDirectoryName(JournalPath);
+
+        /// <summary>
+        /// Creates a journal for a given process.
+        /// 
+        /// Shared opaque directory outputs of <paramref name="process"/> are used as root directories and
+        /// <see cref="Pip.FormattedSemiStableHash"/> is used as journal base name.
+        ///
+        /// <seealso cref="SharedOpaqueJournal(PathTable, IEnumerable{AbsolutePath}, AbsolutePath)"/>
+        /// </summary>
+        public SharedOpaqueJournal(PipExecutionContext context, Process process, AbsolutePath journalDirectory)
+            : this(
+                  context.PathTable, 
+                  process.DirectoryOutputs.Where(d => d.IsSharedOpaque).Select(d => d.Path),
+                  GetJournalFileForProcess(context.PathTable, journalDirectory, process))
+        {
+            Contract.Requires(process != null);
+            Contract.Requires(context != null);
+            Contract.Requires(journalDirectory.IsValid);
+            Contract.Requires(Directory.Exists(journalDirectory.ToString(context.PathTable)));
+        }
+
+        /// <summary>
+        /// Creates a new journal.
+        /// </summary>
+        /// <param name="pathTable">Path table.</param>
+        /// <param name="rootDirectories">Only paths under one of the root directories are recorded in <see cref="RecordFileWrite(AbsolutePath)"/>.</param>
+        /// <param name="journalPath">File to which to write recorded accesses.</param>
+        public SharedOpaqueJournal(PathTable pathTable, IEnumerable<AbsolutePath> rootDirectories, AbsolutePath journalPath)
+        {
+            Contract.Requires(pathTable.IsValid);
+            Contract.Requires(rootDirectories != null);
+
+            m_pathTable = pathTable;
+            m_rootDirectories = rootDirectories.ToArray();
+
+            JournalPath = journalPath.ToString(pathTable);
+            m_bxlWriter = new BuildXLWriter(
+                stream: new FileStream(JournalPath, FileMode.Create, FileAccess.Write, FileShare.Read | FileShare.Delete),
+                debug: false,
+                logStats: false,
+                leaveOpen: false);
+        }
+
+        /// <summary>
+        /// Given a root directory (<paramref name="journalDirectory"/>), returns the full path to the journal corresponding to process <paramref name="process"/>.
+        /// </summary>
+        public static AbsolutePath GetJournalFileForProcess(PathTable pathTable, AbsolutePath journalDirectory, Process process)
+        {
+            Contract.Requires(journalDirectory.IsValid);
+            return journalDirectory.Combine(pathTable, process.FormattedSemiStableHash);
+        }
+
+        /// <summary>
+        /// Finds and returns all journal files that exist in directory denoted by <paramref name="directory"/>
+        /// </summary>
+        public static IEnumerable<string> FindAllJournalFiles(string directory)

```

Comment:
>FindAllJournalFiles [](start = 42, length = 19)

Another consideration is that the number of files in the same journal dir folder can grow to be a very big number, given that there is a 1:1 correspondence to the number of pips. In similar cases in the past we 'clustered' files in different subdirs, based on some grouping factor.
This is usually to avoid enumerations and other file operations that don't deal well with a huge amount of files in the same dir

Reply:
added subduers